cluster_experiments/perturbator.py

```diff
@@ -350,6 +351,274 @@ def _sample_beta_effect(
+        if range_min >= range_max:
+            raise ValueError(
+                f"range_min needs to be smaller than range_max, got "
+                f"range_min={range_min} and range_max={range_max}"
+            )
+
+    def check_relative_effect_bounds(
+        self, df: pd.DataFrame, average_effect: float
+    ) -> None:
+        self.check_average_effect_greater_than(average_effect, x=self._range_min)
+        self.check_average_effect_smaller_than(average_effect, x=self._range_max)
+        self.check_target_is_not_constant_zero(df, average_effect)
+
+    def check_average_effect_greater_than(
+        self, average_effect: float, x: float
+    ) -> Optional[NoReturn]:
+        if average_effect <= x:
+            raise ValueError(
+                f"Simulated effect needs to be greater than range_min={x}, got {average_effect}"
+            )
+
+    def check_average_effect_smaller_than(
+        self, average_effect: float, x: float
+    ) -> Optional[NoReturn]:
+        if average_effect >= x:
+            raise ValueError(
+                f"Simulated effect needs to be smaller than range_max={x}, got {average_effect}"
+            )
+
+    def _sample_scaled_beta_effect(
+        self, average_effect: float, scale: float, n: int
+    ) -> np.ndarray:
+        average_effect_inv_transf = self._inv_transform_to_range(average_effect)
+        scale_inv_transf = self._inv_transform_to_range(scale)
+        a = average_effect_inv_transf / (scale_inv_transf * scale_inv_transf)
+        b = (1 - average_effect_inv_transf) / (scale_inv_transf * scale_inv_transf)
+
+        # the division by abs(average_effect) makes the distribution more concentrated around the mean
+        beta = np.random.beta(a / abs(average_effect), b / abs(average_effect), n)
+
+        return self._transform_to_range(beta)
+
+    def _transform_to_range(self, x: Union[float, np.ndarray]):
+        return x * (self._range_max - self._range_min) + self._range_min
+
+    def _inv_transform_to_range(self, x: Union[float, np.ndarray]):
+        return (x - self._range_min) / (self._range_max - self._range_min)
+
+    @classmethod
+    def from_config(cls, config):
+        """Creates a Perturbator object from a PowerConfig object"""
+        return cls(
+            average_effect=config.average_effect,
+            target_col=config.target_col,
+            treatment_col=config.treatment_col,
+            treatment=config.treatment,
+            scale=config.scale,
+            range_min=config.range_min,
+            range_max=config.range_max,
+        )
+
+
+class ClusteredBetaRelativePerturbator(BetaRelativePositivePerturbator):
+    """
+    A stochastic Perturbator for continuous targets that applies a sampled
+    effect from the Beta distribution. It applies the effect multiplicatively
+    and based on given clusters.
+    For each cluster, the average cluster effect is sampled from a beta
```

Comment:
is there another name which is not cluster? I think we may have a lot of confusion with clusters from the cluster-randomized trial, in splitter and analysis

Reply:
Do you have a suggestion? I agree that the clustering should be distinguished better on the 3 levels of splitter, perturbator and analysis (e.g. as I've tried to start doing in the power config). But I'm not sure how helpful a new name can be, considering that the purpose is really the same (just on a different level). But I may also just be thinking that because I can't see another option that I like^^

Reply:
The purpose is really the same -> this is what I don't fully get. Both in the splitter and analysis, cluster_cols have a very precise meaning: it is the unit of treatment assignment: if in your experiment design, you assign cities to the treatment, then the clusters should be cities. If you assign stores, it should be stores.

In this case, I don't see how the clusters are related to units of treatment assignment. For instance, I understand you could be applying the treatment to cities and have the perturbation work differently for different stores. 

Do you agree? if so, we can think of a better name

Reply:
Good point, I agree. I guess I thought about it too much in terms of the backtests case, where most likely you'd like the clusters in both the perturbation and the analysis to be the same (clusters in the splitter don't matter in the backtest case, cause it's not splitting anything).
So, any suggestion for the naming?^^

Reply:
wdyt about groups, perturbation_groups, GroupedBetaRelativePerturbator, etc?

Reply:
Changing to `Segmented`